base/rational.jl

```diff
@@ -10,42 +27,30 @@ struct Rational{T<:Integer} <: Real
     num::T
     den::T
 
-    function Rational{T}(num::Integer, den::Integer) where T<:Integer
-        num == den == zero(T) && __throw_rational_argerror_zero(T)
-        num2, den2 = divgcd(num, den)
-        if T<:Signed && signbit(den2)
-            den2 = -den2
-            signbit(den2) && __throw_rational_argerror_typemin(T)
-            num2 = -num2
-        end
-        return new(num2, den2)
-    end
-
-    function Rational{T}(num::Integer, den::Integer, ::Val{false}) where T<:Integer
-        # Used when num and den are only known to be coprime and not both equal to 0
-        if T<:Signed && signbit(den)
+    function (::Type{unsafe_rational{T}})(num::Integer, den::Integer; checksign::Bool=false) where T<:Integer
```

Comment:
```suggestion
    function (::Type{unsafe_rational{T}})(num::Integer, den::Integer; check_not_typemin::Bool=false) where T<:Integer
```
Although I'm not really sure this option is needed at all

Reply:
I think it's ok not to check for typmin here, only in the safe constructor

Reply:
Actually it's not in the use cases of this file, and I think it's even more general. For instance, for the `inv` function, you may well start with the rational `typemin(Int) // typemax(Int)`, which is valid, but its inverse is not and the check for `typemin` is necessary here, right?

Reply:
Yes, my argument is that because this is marked as unsafe, checking this should be the responsibility of the caller, i.e. `inv` for example. The check should be pretty cheap though, so I'm not so sure anymore, that there will be a real performance advantage.

Reply:
Ah, I understand. But I still believe that in all cases where a sign check is needed, a typemin check is actually also needed (except maybe in some really contrived cases for which the user may as well use `checksign=false` and do their own check).
I'll gladly look at a counter-example if you have one, that would be interesting!

Reply:
Ah, I understand. But I still believe that in all cases where a sign check is needed, a typemin check is actually also needed (except maybe in some really contrived cases for which the user may as well use `checksign=false` and do their own check).
I'll gladly look at a counter-example if you have one, that would be interesting!

Reply:
The only problematic case is if the user passes in `typemin(Int)` as denominator, in which case `-typemin(Int) == typemin(Int)`, which is what is checked here. If the caller guarantees that `den != typemin(Int)`, then everything will be fine. Negating a positive number will never produce `typemin(Int)` for two's complement integers.

Reply:
> Negating a positive number will never produce `typemin(Int)` for two's complement integers.

Ah, that was exactly the counterexample that eluded me! Very good point.

Reply:
The choice is quite difficult to make. Here are some pros and cons for keeping the current check:

PROS:
- Not very costly and safer.
- If the check is known to be useless, the user may switch to `checksign=false` and do the specific sign check on their own.
- All cases of `checksign=true` in this file actually need this check.

CONS:
- Needlessly safe since the operation is already marked `unsafe`. If the user needs this additional check they can do it themselves
- A useless check is performance lost.

I still prefer to keep it this way because I'm still somewhat convinced that most use cases actually need it. The case where you negate a positive number and give it as a denominator is not such a good counterexample because either you know the sign of your number, and then you should use `checksign=false` with the appropriate signs for the numerator and denominator (just negate the numerator instead of the denominator), either you don't know its sign and then the check is necessary in case you just negated `typemin`.

But I'm ready to be swayed by another argument or if someone has a strong feeling it should be the other way round! The question deserves consideration.

Reply:
Seems reasonable to me